From 349a7500a230b3fc0923e1e6f25309cc40193d8b Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Wed, 22 Jul 2020 22:24:12 +1000
Subject: [PATCH] use from


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index c018f9d94..72f06e004 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -270,9 +270,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                     this.lastPitch = to.getPitch();
 
                     if (to.distance(from) > 0 && !getPlayer().canMove() ){//hard stop.
-                        this.player.playerConnection.sendPacket(new PacketPlayOutPosition(d0, d1, d2,
-                                to.getYaw(),
-                                to.getPitch(), Collections.emptySet()));
+                        this.player.playerConnection.sendPacket(new PacketPlayOutPosition(from.getX(), from.getY(), from.getZ(), from.getYaw(), from.getPitch(), Collections.<PacketPlayOutPosition.EnumPlayerTeleportFlags>emptySet()));
                         return;
                     }
 
-- 
2.24.1.windows.2

