From 7fce232750b48a696d2b132c3adafb3f4d0eef8b Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Wed, 22 Jul 2020 22:29:23 +1000
Subject: [PATCH] hmm


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 72f06e004..7b4498001 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -269,8 +269,8 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                     this.lastYaw = to.getYaw();
                     this.lastPitch = to.getPitch();
 
-                    if (to.distance(from) > 0 && !getPlayer().canMove() ){//hard stop.
-                        this.player.playerConnection.sendPacket(new PacketPlayOutPosition(from.getX(), from.getY(), from.getZ(), from.getYaw(), from.getPitch(), Collections.<PacketPlayOutPosition.EnumPlayerTeleportFlags>emptySet()));
+                    if (to.distance(from) > 0 && !getPlayer().canMove()){//hard stop.
+                        this.player.playerConnection.sendPacket(new PacketPlayOutPosition(from.getX(), from.getY(), from.getZ(), to.getYaw(), to.getPitch(), Collections.<PacketPlayOutPosition.EnumPlayerTeleportFlags>emptySet()));
                         return;
                     }
 
-- 
2.24.1.windows.2

