From eab701eb4c587d1bb439f03cfc7b3c5212ba3403 Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Sat, 8 Aug 2020 01:23:52 +1000
Subject: [PATCH] run schedules every x


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 07718760b..ebfd58a72 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -788,7 +788,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
 
         SpigotTimings.bukkitSchedulerTimer.startTiming(); // Spigot
         // CraftBukkit start
-        this.server.getScheduler().mainThreadHeartbeat(this.ticks);
+        if (this.ticks % getMultiplier() == 0) this.server.getScheduler().mainThreadHeartbeat(this.ticks);
         SpigotTimings.bukkitSchedulerTimer.stopTiming(); // Spigot
 
         // Run tasks that are waiting on processing
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
index ac322824c..b7a0f76dc 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -396,8 +396,8 @@ public class CraftScheduler implements BukkitScheduler {
     }
 
     private CraftTask handle(final CraftTask task, long delay) {
-        if (delay > 0) delay = delay * (MinecraftServer.getMultiplier());
-        task.setPeriod(task.getPeriod() * (MinecraftServer.getMultiplier()));
+        //if (delay > 0) delay = delay * (MinecraftServer.getMultiplier());
+        //task.setPeriod(task.getPeriod() * (MinecraftServer.getMultiplier()));
         task.setNextRun(currentTick + delay);
         addTask(task);
         return task;
-- 
2.24.1.windows.2

