From 656f21f7314076cec375f518dcf7941a93d52f83 Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Sat, 8 Aug 2020 01:52:08 +1000
Subject: [PATCH] return actual period < 0


diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
index 4b1e35236..7c89c6d1a 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
@@ -61,7 +61,7 @@ public class CraftTask implements BukkitTask, Runnable { // Spigot
     }
 
     long getPeriod() {
-        return period;
+        return period > 0 ? period * MinecraftServer.getMultiplier() : period;
     }
 
     void setPeriod(long period) {
-- 
2.24.1.windows.2

