From 9a9fbff2c57774aa8b1d90bbf7ee49a23728830c Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Sat, 27 Jun 2020 18:18:53 +1000
Subject: [PATCH] Expose the sound methods


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 4fddece8..53b13e8b 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -7,6 +7,7 @@ import java.util.Set;
 
 import org.bukkit.Location;
 import org.bukkit.Material;
+import org.bukkit.Sound;
 import org.bukkit.block.Block;
 import org.bukkit.inventory.EntityEquipment;
 import org.bukkit.potion.PotionEffect;
@@ -18,6 +19,25 @@ import org.bukkit.projectiles.ProjectileSource;
  */
 public interface LivingEntity extends Entity, Damageable, ProjectileSource {
 
+    public boolean canMove();
+    public boolean canCollide();
+
+    public void setCanMove(boolean canMove);
+    public void setCanCollide(boolean canCollide);
+
+    public Sound getIdleSound();
+    public Sound getHurtSound();
+    public Sound getDeathSound();
+
+    public void setIdleSound(Sound idleSound);
+    public void setIdleSound(String idleSound);
+
+    public void setHurtSound(Sound hurtSound);
+    public void setHurtSound(String hurtSound);
+
+    public void setDeathSound(Sound deathSound);
+    public void setDeathSound(String deathSound);
+
 
     /**
      * Gets the height of the living entity's eyes above its Location.
-- 
2.24.1.windows.2

