From 213d21aef5efb9adad4ce52cae8d0aece46c6e5f Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Tue, 7 Jul 2020 16:03:50 +1000
Subject: [PATCH] Add LocalizedName Support


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 53b13e8b..fbd202bd 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -29,13 +29,37 @@ public interface LivingEntity extends Entity, Damageable, ProjectileSource {
     public Sound getHurtSound();
     public Sound getDeathSound();
 
+    /**
+     *
+     * @param idleSound The new Idle Sound
+     */
     public void setIdleSound(Sound idleSound);
+    /**
+     *
+     * @param idleSound The new Idle Sound
+     */
     public void setIdleSound(String idleSound);
 
+    /**
+     *
+     * @param hurtSound The new Hurt Sound
+     */
     public void setHurtSound(Sound hurtSound);
+    /**
+     *
+     * @param hurtSound The new Hurt Sound
+     */
     public void setHurtSound(String hurtSound);
 
+    /**
+     *
+     * @param deathSound The new Death Sound
+     */
     public void setDeathSound(Sound deathSound);
+    /**
+     *
+     * @param deathSound The new Death Sound
+     */
     public void setDeathSound(String deathSound);
 
 
diff --git a/src/main/java/org/bukkit/inventory/meta/ItemMeta.java b/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
index d8cc8218..21c7961d 100644
--- a/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
@@ -158,6 +158,66 @@ public interface ItemMeta extends Cloneable, ConfigurationSerializable {
     @SuppressWarnings("javadoc")
     ItemMeta clone();
 
+    /**
+     * Checks for existence of a localized name.
+     *
+     * @return true if this has a localized name
+     */
+    boolean hasLocalizedName();
+
+    /**
+     * Gets the localized display name that is set.
+     * <p>
+     * Plugins should check that hasLocalizedName() returns <code>true</code>
+     * before calling this method.
+     *
+     * @return the localized name that is set
+     */
+     String getLocalizedName();
+
+    /**
+     * Sets the localized name.
+     *
+     * @param name the name to set
+     */
+    void setLocalizedName(String name);
+
+    // Bandaid Start
+    public class BandaidPaper{
+        /**
+         * Checks for existence of a localized name.
+         *
+         * @return true if this has a localized name
+         */
+        public boolean hasLocalizedName(){
+            throw new UnsupportedOperationException( "Not supported yet." );
+        }
+
+        /**
+         * Gets the localized display name that is set.
+         * <p>
+         * Plugins should check that hasLocalizedName() returns <code>true</code>
+         * before calling this method.
+         *
+         * @return the localized name that is set
+         */
+        public String getLocalizedName(){
+            throw new UnsupportedOperationException( "Not supported yet." );
+        }
+
+        /**
+         * Sets the localized name.
+         *
+         * @param name the name to set
+         */
+        public void setLocalizedName(String name){
+            throw new UnsupportedOperationException( "Not supported yet." );
+        }
+    }
+    // Bandaid End
+
+    BandaidPaper bandaid();
+
     // Spigot start
     public class Spigot
     {
-- 
2.24.1.windows.2

