From 3f2e02ab3a22dbad1fe97e934b7d67afc307908c Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Fri, 10 Jul 2020 19:43:22 +1000
Subject: [PATCH] Add references to blocks and items


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 0fb26055..a4310123 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -165,6 +165,7 @@ public enum Material {
     HUGE_MUSHROOM_1(99, Mushroom.class),
     HUGE_MUSHROOM_2(100, Mushroom.class),
     IRON_FENCE(101),
+    IRON_BARS(101),
     THIN_GLASS(102),
     MELON_BLOCK(103),
     PUMPKIN_STEM(104, MaterialData.class),
@@ -174,6 +175,7 @@ public enum Material {
     BRICK_STAIRS(108, Stairs.class),
     SMOOTH_STAIRS(109, Stairs.class),
     MYCEL(110),
+    MYCELIUM(110),
     WATER_LILY(111),
     NETHER_BRICK(112),
     NETHER_FENCE(113),
@@ -296,6 +298,7 @@ public enum Material {
     STRING(287),
     FEATHER(288),
     SULPHUR(289),
+    GUNPOWDER(289),
     WOOD_HOE(290, 1, 59),
     STONE_HOE(291, 1, 131),
     IRON_HOE(292, 1, 250),
@@ -382,6 +385,7 @@ public enum Material {
     GHAST_TEAR(370),
     GOLD_NUGGET(371),
     NETHER_STALK(372),
+    NETHER_WART(372),
     /**
      * @see Potion
      */
@@ -396,6 +400,7 @@ public enum Material {
     EYE_OF_ENDER(381),
     SPECKLED_MELON(382),
     MONSTER_EGG(383, 64, SpawnEgg.class),
+    SPAWN_EGG(383, 64, SpawnEgg.class),
     EXP_BOTTLE(384, 64),
     FIREBALL(385, 64),
     BOOK_AND_QUILL(386, 1),
@@ -773,6 +778,7 @@ public enum Material {
             case HUGE_MUSHROOM_1:
             case HUGE_MUSHROOM_2:
             case IRON_FENCE:
+            case IRON_BARS:
             case THIN_GLASS:
             case MELON_BLOCK:
             case FENCE_GATE:
@@ -1095,6 +1101,7 @@ public enum Material {
             case HUGE_MUSHROOM_2:
             case MELON_BLOCK:
             case MYCEL:
+            case MYCELIUM:
             case NETHER_BRICK:
             case ENDER_PORTAL_FRAME:
             case ENDER_STONE:
-- 
2.24.1.windows.2

