From f97cc7ce0dc4f675932ed33d6bf7e7be39600d26 Mon Sep 17 00:00:00 2001
From: Gatt <gatt@gatt.space>
Date: Sat, 11 Jul 2020 13:11:45 +1000
Subject: [PATCH] walk sounds


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index fbd202bd..c7b6ebf2 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -28,7 +28,20 @@ public interface LivingEntity extends Entity, Damageable, ProjectileSource {
     public Sound getIdleSound();
     public Sound getHurtSound();
     public Sound getDeathSound();
+    public Sound getWalkSound();
+    /**
+     *
+     * @param walkSound The new Idle Sound
+     */
+    public void setWalkSound(Sound walkSound); /**
+     *
+     * @param walkSound The new Idle Sound
+     */
+    public void setWalkSound(String walkSound);
 
+    /**
+     *
+     */
     /**
      *
      * @param idleSound The new Idle Sound
-- 
2.24.1.windows.2

